frame/staking/src/offchain_election.rs

```diff
@@ -0,0 +1,247 @@
+					// value doesn't exists. Probably this node just booted up. Write, and run
+					Ok(now)
+				}
+			}
+		});
+
+	match mutate_stat {
+		// all good
+		Ok(Ok(_)) => Ok(()),
+		// failed to write.
+		Ok(Err(_)) => Err("failed to write to offchain db."),
+		// fork etc.
+		Err(why) => Err(why),
+	}
+}
+
+/// The internal logic of the offchain worker of this module. This runs the phragmen election,
+/// compacts and reduces the solution, computes the score and submits it back to the chain as an
+/// unsigned transaction.
+pub(crate) fn compute_offchain_election<T: Trait>() -> Result<(), OffchainElectionError> {
+	let keys = <Module<T>>::keys();
+	let local_keys = T::KeyType::all();
+
+	// For each local key is in the stored authority keys, try and submit. Breaks out after first
+	// successful submission.
+	for (index, ref pubkey) in local_keys.into_iter().filter_map(|key| {
+		keys.iter()
+			.enumerate()
+			.find(|(_, val_key)| **val_key == key)
+	}) {
+		// compute raw solution. Note that we use `OffchainAccuracy`.
+		let PhragmenResult {
+			winners,
+			assignments,
+		} = <Module<T>>::do_phragmen::<OffchainAccuracy>()
+			.ok_or(OffchainElectionError::ElectionFailed)?;
+
+		// process and prepare it for submission.
+		let (winners, compact, score) = prepare_submission::<T>(assignments, winners, true)?;
+
+		// sign it.
+		let signature_payload: SignaturePayload = (&winners, &compact, &score, &(index as u32));
+		let signature = pubkey
+			.sign(&signature_payload.encode())
+			.ok_or(OffchainElectionError::SigningFailed)?;
+
+		// send it.
+		let call: <T as Trait>::Call = Call::submit_election_solution_unsigned(
+			winners,
+			compact,
+			score,
+			index as u32,
+			signature,
+		).into();
+
+		if T::SubmitTransaction::submit_unsigned(call).map_err(|_| {
+			debug::native::warn!(
+				target: "staking",
+				"failed to submit offchain solution with key {:?}",
+				pubkey,
+			);
+		}).is_ok() {
+			// return and break out after the first successful submission
+			return Ok(());
+		}
+	}
+
+	// no key left and no submission.
+	Err(OffchainElectionError::NoSigningKey)
+}
+
+/// Takes a phragmen result and spits out some data that can be submitted to the chain.
+///
+/// This does a lot of stuff; read the inline comments.
+pub fn prepare_submission<T: Trait>(
+	assignments: Vec<Assignment<T::AccountId, OffchainAccuracy>>,
+	winners: Vec<(T::AccountId, ExtendedBalance)>,
+	do_reduce: bool,
+) -> Result<(Vec<ValidatorIndex>, CompactAssignments, PhragmenScore), OffchainElectionError>
+where
+	ExtendedBalance: From<<OffchainAccuracy as PerThing>::Inner>,
+{
+	// make sure that the snapshot is available.
+	let snapshot_validators =
+		<Module<T>>::snapshot_validators().ok_or(OffchainElectionError::SnapshotUnavailable)?;
+	let snapshot_nominators =
+		<Module<T>>::snapshot_nominators().ok_or(OffchainElectionError::SnapshotUnavailable)?;
+
+	// all helper closures
+	let nominator_index = |a: &T::AccountId| -> Option<NominatorIndex> {
```

Comment:
isn't this a bit slow to do repeatedly?

Reply:
Surely is, but this runs offchain so we should be fine. There are numerous places in the offchain code path that you can optimise (more significant than this one) but I'd leave them for the future.

Reply:
For the future, if we decide to keep the snapshots around, we should sort them and then these searches over them will be cheaper. Also, the other one for fetching the index of the validator could also use an internal cache as it will be called with duplicates.